**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: ARAH ROJAS BLANCO
*   MATRÍCULA: A00834299


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [3]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [4]:
query = sqla.text("""
SELECT *
FROM movie
WHERE runtime > 300
ORDER BY runtime DESC
""")
peliculas_largas = pd.read_sql(query, conn)
peliculas_largas

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
3,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...
4,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series


El runtime esta expresado en minutos por lo que se hace la conversion de 5 horas a minutos, dando como resultado 300 minutos.
Se obtienen cinco registros: Nihontouitsu Series: Film Series (3892 minutos), Free and Easy Series: Film Series (2120), Seiji Oda: Film Series (710), Kingdom ~ The Man Who Became the Top ~: Film Series (427) y Navarasa: Limited Series (312). Sus títulos indican que hay agrupaciones de películas o series dentro de movie; se reportan tal como están almacenadas.

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [5]:
query = sqla.text("""
SELECT ROUND(
    100.0 * SUM(available_globally = 0) / COUNT(available_globally),
    2
) AS porcentaje_solo_eu
FROM movie
""")
porcentaje_eu = pd.read_sql(query, conn)
porcentaje_eu

,porcentaje_solo_eu
0,83.57


Resultado: el 83.57 % de las películas con disponibilidad conocida están disponibles únicamente en Estados Unidos, de acuerdo con ese indicador.

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [6]:
query = sqla.text("""
SELECT DISTINCT locale
FROM movie
ORDER BY locale
""")
idiomas = pd.read_sql(query, conn)
idiomas

,locale
0,None
1,en


In [7]:
query = sqla.text("""
SELECT COUNT(*) AS cantidad_locale_nulo
FROM movie
WHERE locale IS NULL
""")
locale_nulos = pd.read_sql(query, conn)
locale_nulos

,cantidad_locale_nulo
0,11255


El único código de idioma registrado es en (inglés); también aparece NULL, que representa información ausente, no un idioma o región. Hay 11 255 películas con locale nulo. En SQL los nulos se identifican con IS NULL.

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [8]:
query = sqla.text("""
SELECT
    SUM(title <> original_title) AS titulos_diferentes,
    SUM(locale IS NULL AND original_title IS NULL) AS sin_titulo_original,
    SUM(locale IS NULL AND title = original_title) AS titulos_equivalentes
FROM movie
""")
comparacion_titulos = pd.read_sql(query, conn)
comparacion_titulos

,titulos_diferentes,sin_titulo_original,titulos_equivalentes
0,3947.0,7291.0,17.0


Hay 3947 películas con títulos diferentes según la comparación de MySQL. No coincide con los 11 255 registros con locale nulo de la pregunta 3. Los 7308 restantes tienen el título original ausente (7291) o títulos que MySQL considera equivalentes (17).
MySQL ignora ciertas diferencias de acentos y mayúsculas al comparar estos campos; por ejemplo, considera equivalentes Animas y Ánimas.
Conclusión: no es válido asumir que locale = NULL significa un idioma distinto del inglés. Representa un dato desconocido, y la comparación de títulos no permite determinar el idioma.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

cumulative_weeks_in_top10 ya acumula las semanas. Por ello se toma el máximo de cada película en los resúmenes semanales; sumarlo duplicaría semanas. HAVING conserva todas las películas que empaten con el mayor valor registrado.

In [9]:
query = sqla.text("""
SELECT m.title,
       MAX(v.cumulative_weeks_in_top10) AS semanas_en_top10
FROM movie AS m
JOIN view_summary AS v ON v.movie_id = m.id
WHERE v.duration = 'WEEKLY'
GROUP BY m.id, m.title
HAVING MAX(v.cumulative_weeks_in_top10) = (
    SELECT MAX(cumulative_weeks_in_top10)
    FROM view_summary
    WHERE movie_id IS NOT NULL
      AND duration = 'WEEKLY'
)
ORDER BY m.title
""")
pelicula_top10 = pd.read_sql(query, conn)
pelicula_top10

,title,semanas_en_top10
0,The Boss Baby,22


Resultado: The Boss Baby tiene la mayor permanencia registrada, con 22 semanas acumuladas en el top 10. No implica que esas semanas hayan sido consecutivas.

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

Se relacionan los programas con sus temporadas, se cuentan los registros de temporada por programa y se muestran los cinco mayores. El título sirve para ordenar los empates de forma reproducible.

In [10]:
query = sqla.text("""
SELECT t.title, COUNT(s.id) AS cantidad_temporadas
FROM tv_show AS t
JOIN season AS s ON s.tv_show_id = t.id
GROUP BY t.id, t.title
ORDER BY cantidad_temporadas DESC, t.title ASC
LIMIT 5
""")
programas_temporadas = pd.read_sql(query, conn)
programas_temporadas

,title,cantidad_temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


Grey’s Anatomy y Naruto Shippuden tienen 20 temporadas cada uno; Heartland (2007), 17; It’s Always Sunny in Philadelphia, 16; y NCIS, 15. Son las temporadas registradas en esta base, no necesariamente todas las existentes en otros catálogos.

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

El valor que identifica los resúmenes semestrales es SEMI_ANNUALLY. DISTINCT evita repetir el mismo intervalo por cada película o temporada.

In [11]:
query = sqla.text("""
SELECT DISTINCT start_date, end_date
FROM view_summary
WHERE duration = 'SEMI_ANNUALLY'
ORDER BY start_date
""")
periodos_semestrales = pd.read_sql(query, conn)
periodos_semestrales

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


Los intervalos son del 1 de julio al 31 de diciembre de 2023 y del 1 de enero al 30 de junio de 2024, ambos con las fechas extremas incluidas.

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [12]:
periodo_2024 = periodos_semestrales.loc[
    pd.to_datetime(periodos_semestrales['start_date']).eq(pd.Timestamp('2024-01-01'))
].iloc[0]

inicio_2024 = periodo_2024['start_date']
fin_2024 = periodo_2024['end_date']

query = sqla.text("""
SELECT s.season_number, s.title AS temporada, v.views
FROM tv_show AS t
JOIN season AS s ON s.tv_show_id = t.id
JOIN view_summary AS v ON v.season_id = s.id
WHERE t.title = :programa
  AND v.duration = 'SEMI_ANNUALLY'
  AND v.start_date = :inicio
  AND v.end_date = :fin
ORDER BY v.views DESC, s.season_number ASC
""")
greys_sql = pd.read_sql(
    query,
    conn,
    params={
        'programa': "Grey's Anatomy",
        'inicio': inicio_2024,
        'fin': fin_2024
    }
)
greys_sql

,season_number,temporada,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


La temporada 1 encabeza la lista con 3.6 millones de vistas, seguida por la 2 con 3.1 millones. Las temporadas 3, 4 y 5 empatan con 2.9 millones; se ordenan por número de temporada para resolver el empate. La temporada 20 tiene el menor valor, 100 000 vistas.
Se observa una tendencia general a menos vistas en las temporadas posteriores, pero no una disminución estricta: la temporada 19 alcanza 2 millones, por encima de las temporadas 15 a 18. El resultado describe el consumo registrado durante este semestre; no demuestra que las mismas personas abandonaran la serie ni que una temporada tenga mayor calidad. La disponibilidad y el tiempo de exposición podrían influir, pero esta consulta no permite confirmar esas causas.
La métrica de Netflix vistas se obtiene de las horas vistas divididas entre la duración del contenido. No equivale al número de personas distintas ni garantiza reproducciones completas.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [13]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

Se filtran el programa y los resúmenes del mismo intervalo de la pregunta 8. merge() equivale a los JOIN, la selección de columnas equivale al SELECT y sort_values() reproduce el ORDER BY. Se renombran las claves antes de unir para evitar confundir los distintos campos id.

Al cargar toda la tabla season, los números de temporada se representan como decimales porque existen nulos en otros programas. Después de filtrar Grey’s Anatomy, cuyos números están completos, se convierten a enteros para conservar el mismo tipo de dato que en la consulta SQL

In [14]:
programa = tv_show.loc[
    tv_show['title'].eq("Grey's Anatomy"), ['id']
].rename(columns={'id': 'tv_show_id'})

temporadas = season[
    ['id', 'tv_show_id', 'season_number', 'title']
].rename(columns={'id': 'season_id', 'title': 'temporada'})

resumen = view_summary.loc[
    view_summary['duration'].eq('SEMI_ANNUALLY')
    & view_summary['start_date'].eq(inicio_2024)
    & view_summary['end_date'].eq(fin_2024),
    ['season_id', 'views']
]

greys_pandas = (
    programa
    .merge(temporadas, on='tv_show_id')
    .merge(resumen, on='season_id')
    [['season_number', 'temporada', 'views']]
    .astype({'season_number': 'int64'})
    .sort_values(['views', 'season_number'], ascending=[False, True])
    .reset_index(drop=True)
)
greys_pandas

,season_number,temporada,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


In [15]:
greys_pandas.equals(greys_sql)

True

Pandas devuelve las mismas 20 temporadas, vistas y orden que SQL. La comparación devuelve True, por lo que también aplica la interpretación de la pregunta 8.

`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [16]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [17]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [18]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Act3DB"
os.chdir(DIR)

In [19]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [20]:
db_firestore = firestore.client()

In [21]:
documentos = db_firestore.collection('EMPLOYEE').stream()
registros = []

for doc in documentos:
    empleado = doc.to_dict()
    empleado['id'] = doc.id
    registros.append(empleado)

empleados = pd.DataFrame(
    registros,
    columns=['id', 'emp_fname', 'emp_lname', 'emp_hiredate']
).set_index('id')
empleados

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


Se recuperan cuatro empleados: David Senior, William Smithfield, John News y June Arbough, junto con sus fechas de contratación. El identificador de cada documento aparece como índice id, tal como solicita el ejemplo; las fechas conservan su zona horaria UTC.

In [22]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---

* Declaración de uso de IA

Se utilizó OpenAI Codex, basado en GPT-6, para apoyar la generación y revisión del código SQL y Pandas y la lectura de Firestore .